### PR TITLE
Improve slider ball and follow circle animations

### PR DESCRIPTION
**Pull request by goodtrailer**
Attempts to more accurately match some of stable's animations (#13660), specifically slider ball and follow circle fade out. On stable, slider balls disappear instantly, and follow circles fade out way slower.
~~Also, stable's follow circles are noticeably smaller, though this part of the PR I'm less sure about because it changes follow circles from 2.4x to 2.0x size, actually affecting gameplay (I think?). I'm not even sure if 2.0x is the right number, since follow circles on stable do a bouncing animation.~~

Comparison video for legacy skins:
~~https://www.youtube.com/watch?v=tNnfiz4EOG4~~
https://www.youtube.com/watch?v=sn6oaPXcbkM

Comparison video for default skin (which should not be changed by this PR):
https://www.youtube.com/watch?v=T6cHvewvkcI

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs [hunk 1]
```diff
@@ -103,14 +103,22 @@ private set
...
+                        followCircle.ScaleTo(1.6f, 200, Easing.None);
+                }
                 else
                 {
                     // We need to always be tracking the final size, at both endpoints. For now, this is achieved by removing the scale duration.
-                    followCircle.ScaleTo(tracking ? 2.4f : 1f);
+                    followCircle.ScaleTo(tracking ? 2f : 1f);
                 }
 
-                followCircle.FadeTo(tracking ? 1f : 0, 300, Easing.OutQuint);
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs:113]: you're changing the defaults. i think you intended to only change legacy skins.

File: osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs [hunk 2]
```diff
@@ -103,14 +103,22 @@ private set
                 tracking = value;
 
                 if (InputTracksVisualSize)
-                    followCircle.ScaleTo(tracking ? 2.4f : 1f, 300, Easing.OutQuint);
+                {
+                    if (tracking)
+                        followCircle.ScaleTo(2f, 200, Easing.OutQuint);
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs:108]: this makes skins track (as per stable's weirndess) an actually incorrect visual area which was intended to be improved in lazer. you're undoing this improvement.

Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs]: Ah, makes sense. Reverted to 2.4x size.

File: osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs (hunk 1 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs]: I may be misunderstanding, but every single `DrawableSlider` instantiates this specific `SliderBall` class. So I'm not sure why `SliderBall` is in Defaults anyway since it's always used, and I don't see a way to only affect legacy skins.

https://github.com/ppy/osu/blob/512af8809ed9a94f98338593472b1ead628e9212/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs#L32
https://github.com/ppy/osu/blob/512af8809ed9a94f98338593472b1ead628e9212/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs#L76-L82

And the legacy slider ball class is super empty:
https://github.com/ppy/osu/blob/512af8809ed9a94f98338593472b1ead628e9212/osu.Game.Rulesets.Osu/Skinning/Legacy/LegacySliderBall.cs

Shouldn't all this `SliderBall` functionality be inside a `DrawableSliderBall` class (i.e. like `DrawableSliderHead`)?

Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs]: You'll have to refactor to allow this. There's plenty of examples.

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            using (BeginAbsoluteSequence(slider.HitObject.StartTime))
+                this.FadeIn().ScaleTo(1f);
+
+            using (BeginAbsoluteSequence(slider.HitObject.EndTime))
+            {
+                this.FadeOut(fade_out_time, Easing.Out);
+                switch (state)
+                {
+                    case ArmedState.Hit:
+                        this.ScaleTo(1.1f, fade_out_time, Easing.Out);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs:77]: Is this supposed to be the scale transform that was in `DrawableSlider`? If so, why has the final value changed (from `HitObject.Scale * 1.4f`)?

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderBall.cs [hunk 4]
```diff
@@ -1,64 +1,65 @@
-﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
-// See the LICENCE file in the repository root for full licence text.
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderBall.cs:2]: License header has been removed. Needs to be reinstated.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderBall.cs [hunk 5]
```diff
@@ -207,73 +186,35 @@ public void UpdateProgress(double completionProgress)
             lastPosition = newPos;
         }
 
-        private class FollowCircleContainer : CircularContainer
+        private class FollowReceptor : CircularContainer
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderBall.cs:189]: What was the reasoning for this `FollowReceptor` change? It seems unrelated to the actual goal of this pull.

(I also slightly dislike that it scrapes out the slider's ball the way it does, rather than the slider caching the ball explicitly and this component resolving it explicitly, but I suppose it's not *horrible*.)

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultFollowCircle.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        private DrawableSlider slider;
+
+        [BackgroundDependencyLoader]
+        private void load(DrawableHitObject drawableObject)
+        {
+            slider = (DrawableSlider)drawableObject;
+
+            RelativeSizeAxes = Axes.Both;
+            Alpha = 1f;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/DefaultFollowCircle.cs:25]: shouldn't be required

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultFollowCircle.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                Child = new Box
+                {
+                    Colour = Color4.Orange,
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0.2f,
+                }
+            };
+
+            slider.Tracking.BindValueChanged(trackingChanged, true);
+            slider.ApplyCustomUpdateState += updateStateTransforms;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/DefaultFollowCircle.cs:43]: These subscriptions are never removed, which I suspect is causing a delegate leak (they need to be removed on disposal/skin change). This goes for all four skinnables added here.

Generally binding to other bindables that a class does not own is frowned upon. The pattern to follow there is to create a local bindable and bind to the external one.

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultFollowCircle.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        private void updateStateTransforms(DrawableHitObject obj, ArmedState state)
+        {
+            if (!(obj is DrawableSlider))
+                return;
+            
+            const float fade_out_time = 112.5f;
+
+            using (BeginAbsoluteSequence(slider.HitObject.EndTime))
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/DefaultFollowCircle.cs:68]: I want to say that this should be using `HitStateUpdateTime` but that needs a double-checking.

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            box.FadeTo(tracking.NewValue ? 0.3f : 0.05f, 200, Easing.OutQuint);
+
+        private void updateStateTransforms(DrawableHitObject obj, ArmedState state)
+        {
+            if (!(obj is DrawableSlider))
+                return;
+
+            const float fade_out_time = 112.5f;
+
+            using (BeginAbsoluteSequence(slider.HitObject.StartTime))
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs:68]: Similarly, `StartTime` -> `StateUpdateTime`

File: osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+            bool tracking = e.NewValue;
+
+            if (slider.Ball.InputTracksVisualSize)
+            {
+                if (tracking)
+                    this.ScaleTo(DrawableSliderBall.FOLLOW_AREA, 200, Easing.OutQuint);
+                else
+                    this.ScaleTo(DrawableSliderBall.FOLLOW_AREA * 2, 100, Easing.OutQuint).Then().ScaleTo(1f);
+            }
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs:52]: This is weird. The logic is conditioned upon `InputTracksVisualSize`. But the `DrawableSliderBall`'s input receptor does not scale in the same manner. So am I to understand input will actually *not* track visual size on legacy skins?

I almost think that skinnable components handling `InputTracksVisualSize` is wrong and it's actually `DrawableSliderBall` that should be figuring that logic out.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
Aside from the above review please run inspectcode locally as there are two inspections to fix.

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs (hunk 3 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs:77]: The `fade_out_time` is 1/4 of what it used to be. so the scale to animation is also 1/4 of what it was (10% instead of 40%). Before, `fade_out_time` was misleading because the fade out animation was applied with `fade_out_time/4`, while the full `fade_out_time` was only used for the scale out animation. So really, the old `fade_out_time` was more like a scale out time.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderBall.cs (hunk 5 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderBall.cs:189]: So `FollowReceptor` is separate from the actual follow circles because this pull request is supposed to only affect the animations of legacy skins. But because the old follow areas were tied directly to the follow circles themselves, that would mean legacy skins (with their different animations) would have different follow areas than the default skin. `FollowReceptor` has the same animations as the default skin, so they match perfectly. Legacy skins will be a little off, but that's inevitable if legacy skins and default skin are going to have different animations.

File: osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs (hunk 10 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs:52]: (Might be misunderstanding the question?) Yeah, legacy skins won't match the actual follow area. In stable, there's an animation were the follow circle goes crazy huge when you let go, but I don't think the follow area scales outwards like that too (correct me if I'm wrong). Default/legacy skin will have different animations, and I don't think it makes sense to give them differently animated follow receptors. One of the two skins is going to have slightly unrepresentative animations. It's weird, but I couldn't think of a better solution.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs:53]: if the goal is to make animations work 1:1 (as opposed to actually improving the gameplay readability as was done before this PR), the legacy skin shouldn't be checking this variable at all.

i'm still not sure and not going to review this PR until i am, though.

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs (hunk 3 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs:79]: `Easing.Out` [is not a linear easing function](https://github.com/ppy/osu-framework/blob/269c6c63625e56ec269dcf7024fb0e95baf68d36/osu.Framework/Graphics/Transforms/DefaultEasingFunction.cs#L50-L52). So you can't quarter the final value and the duration and expect to get the same result.

Please just restore these values to what they were originally.

File: osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs (hunk 10 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs:53]: Given the explanation above I'm actually starting to (personal opinion follows) slightly oppose bringing back the legacy follow circle animation at *all*. It just complicates things needlessly and brings back some of the weird obscure mechanics that stable has, that in a nice world have no right to exist but people also got used to.

As things are right now on `master`, there are two simple options: either the follow area is exactly as you see it, or it is a constant size (equal to the one at end time). This pull makes it so that is the case kinda, only on the default skin, and on legacy skins the actual follow circle never follows the animation, but is either smaller than it and *still*  changing size in a manner invisible to the player, or constant size. Try explaining that one to a beginner player, it just sounds insane.

File: osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs (hunk 3 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/Default/DefaultSliderBall.cs]: Derp, I'm dumb. Restored.

File: osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs (hunk 10 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs:53]: That's pretty fair, but at the very least the follow circle fade out animation that plays at the end of a slider (not the one when the player lets go early) should follow old legacy animations, since they come after the slider has been judged so it doesn't matter if they match the follow area, right?

Comment (bdach) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyFollowCircle.cs:53]: Sure, I could agree with that part of it.

**Comment by goodtrailer** [comment]
Currently have unmerged changes ([diff](https://github.com/goodtrailer/osu/compare/improve-slider-anims...goodtrailer:improve-tick-anims)) that basically go all-in on making legacy followcircle anims 1:1 with stable. Not sure if they should be merged here because this PR was originally meant to focus on fade out anims. Also, there's still that unresolved dilemma (https://github.com/ppy/osu/pull/14832#discussion_r715986777) of how closely legacy anims should match stable, as the more they're 1:1, the less accurately the followcircle represents true follow area while those anims are playing.

Changes in the unmerged branch:
* followcircle "pulse" anim on slidertick/repeat
* followcircle "explode" (break) anim only occurs when a slidertick/repeat/tail is actually missed, instead of when the player lets go
* speeds up followcircle "expand" (click) anim on very short sliders so that it finishes playing before the slider ends
* tweaks to followcircle anim easings/durations (i.e. slower "explode", faster "expand", etc.)

Comparison video with unmerged changes:
https://www.youtube.com/watch?v=XdVH0dCwCNo
